File: sdk/tracechain/tracing.py
"""
Context propagation — stores run_id and step_id in Python contextvars so they
flow automatically down the call stack without the user passing them explicitly.
Works correctly across threads because each thread gets its own context copy.
"""
from contextvars import ContextVar
from typing import Optional

_current_run_id:  ContextVar[Optional[str]] = ContextVar("tracechain_run_id",  default=None)
_current_step_id: ContextVar[Optional[str]] = ContextVar("tracechain_step_id", default=None)

# accumulated cost + tokens for the current run (reset per workflow invocation)
_run_total_cost:   ContextVar[float] = ContextVar("tracechain_run_cost",   default=0.0)
_run_total_tokens: ContextVar[int]   = ContextVar("tracechain_run_tokens", default=0)


def get_run_id()  -> Optional[str]: return _current_run_id.get()
def get_step_id() -> Optional[str]: return _current_step_id.get()

def set_run_id(run_id: str):   return _current_run_id.set(run_id)
def set_step_id(step_id: str): return _current_step_id.set(step_id)

def reset_run_id(token):   _current_run_id.reset(token)
def reset_step_id(token):  _current_step_id.reset(token)

def get_run_totals() -> tuple[float, int]:
    return _run_total_cost.get(), _run_total_tokens.get()

def add_llm_usage(cost: float, tokens: int) -> None:
    _run_total_cost.set(_run_total_cost.get() + cost)
    _run_total_tokens.set(_run_total_tokens.get() + tokens)

def reset_run_totals() -> None:
    _run_total_cost.set(0.0)
    _run_total_tokens.set(0)

File: sdk-js/src/tracing.ts
/**
 * Context propagation via AsyncLocalStorage — the Node.js equivalent of
 * Python's contextvars. Each async call chain (workflow invocation) gets
 * its own isolated context, so concurrent workflows never bleed run IDs.
 */
import { AsyncLocalStorage } from "node:async_hooks";

interface TraceContext {
  runId: string | null;
  stepId: string | null;
  totalCost: number;
  totalTokens: number;
}

const _store = new AsyncLocalStorage<TraceContext>();

function _ctx(): TraceContext {
  return _store.getStore() ?? { runId: null, stepId: null, totalCost: 0, totalTokens: 0 };
}

export function getRunId(): string | null  { return _ctx().runId;  }
export function getStepId(): string | null { return _ctx().stepId; }

/** Run `fn` inside a fresh context with the given runId. */
export function withRunId<T>(runId: string, fn: () => T): T {
  const ctx: TraceContext = { runId, stepId: null, totalCost: 0, totalTokens: 0 };
  return _store.run(ctx, fn);
}

export function setStepId(stepId: string | null): void {
  const ctx = _store.getStore();
  if (ctx) ctx.stepId = stepId;
}

export function addLlmUsage(cost: number, tokens: number): void {
  const ctx = _store.getStore();
  if (ctx) { ctx.totalCost += cost; ctx.totalTokens += tokens; }
}

export function getRunTotals(): { cost: number; tokens: number } {
  const ctx = _ctx();
  return { cost: ctx.totalCost, tokens: ctx.totalTokens };
}
